### DISCUSSION TITLE
Switch between language input source on the keyboard

### DISCUSSION THREAD

**Discussion by mayost** [Xpra-org/xpra#4134] [Q&A]
Hello

My server is: Fedora 39
XPRA version: 5.05

Client: Fedora 39
GNOME desktop
With the option to switch between an English or Hebrew keyboard input source

When I connect to the server through the client and let's say I open a text document when I write with English it is fine when I change to Hebrew by (Super L + space)
So when I try to write it does nothing
I also tried with other languages and it doesn't work, only English works
Is there a way to solve this problem?

It should be noted that if I connect to the server like this:
http://192.168.10.86:10000/connect.html

If I choose a login screen
In the section of: Advanced options
Keyboard Layout: Israel
It only writes to me in Hebrew and when I try to change the language using the Super L + space key combination
It doesn't write anything

If I choose a login screen
In the section of: Advanced options
Keyboard Layout: English USA
It only writes to me in English and when I try to change language using the Super L + space key combination
It doesn't write anything

The problem is that he is not ready to accept the mixing of two or more languages

Thank you
![english](https://github.com/Xpra-org/xpra/assets/1280195/9d32c5c1-c724-4bc9-b4cc-12ff9bb8f551)
![hebrew](https://github.com/Xpra-org/xpra/assets/1280195/08014394-99ec-4feb-98a6-734465aec911)

**Comment by totaam** [comment]
There may be more information in your javascript debug console if you enable "keyboard" debugging.
> It only writes to me in English and when I try to change language using the Super L + space key combination

This event is not exposed to the html5 client, so it has no way of knowing that it needs to change the keymap.

My guess is that we probably need to change the language selection box to allow multiple languages to be selected.

**Reply by mayost** [reply]
1. If I turn on keyboard debugging
Where do I see the file where the debugging is recorded

2. I changed the language change to a combination of other keys (without the SUPER button) now it does show me in the desktop that it changes language but in practice:
It doesn't write Hebrew if I chose keyboard layout: ENGLISH when logging in
It doesn't write English if at the entrance I chose keyboard layout: ISRAEL

I am attaching the log file:
/run/user/1000/xpra/0/server.log
  There is some comment about the keyboard, but I didn't understand what should be done

Thank you
[server.log](https://github.com/Xpra-org/xpra/files/14276309/server.log)

**Reply by totaam** [reply]
> > There may be more information in your javascript debug console if you enable "keyboard" debugging.

> Where do I see the file where the debugging is recorded

In the javascript debug console.

> I changed the language change to a combination of other keys..

What matters more is what the browsers sends, not just the layout that you select in your xpra desktop session.

**Reply by mayost** [reply]
Hello

On the server I ran:
`xpra start --start=xterm --bind-tcp=0.0.0.0:10000`

I connected to the server via HTML5
In the browser I see the XTERM
in the XTERM client
I ran the command:
`setxkbmap -layout us,il ,, -option grp:alt_shift_toggle`

The output of: `setxkbmap -query`
```
rules:      evdev
model:    evdev
layout:    us, il
variant:   ,,
options:  grp:alt_shift_toggle
```
When I try to change keyboard layout by the keys ALT+SHIFT
It does nothing and leaves the keyboard on English

But if I run this command:
`setxkbmap -layout il`

It changes the keyboard to Hebrew and I can write characters in Hebrew
But it is not possible to write English until I run the command again with:
  `setxkbmap -layout us`

Is there a solution to this problem so that it is possible to change keyboard layout languages by pressing ALT+CTRL or any other key combination

Thank you

**Reply by mayost** [reply]
Hello

I am sharing more things that might help to solve the problem:

Server:
1. I installed a new server with FEDORA 39 with minimal installation without desktop and without X11
2. I installed XPRA by dnf install xpra (without the XPRA repositories) it installed version 5.0.2

client:
I installed FEDORA 39 with GNOME desktop
I added another Hebrew keyboard layout (other languages are also possible)
In the local client when I run:
`setxkbmap -query`

This gives a result:
```
rules: evdev
model: pc105+inet
layout: us,il,us
variant: ,,
options: terminate: ctrl_alt_bksp
```
The default language switch in Fedora is: WIN+SPACE

So far so good

But if I run on the server for example:
`xpra start --start=xterm`

And through the client: (FEDORA 39 with GNOME desktop) I connect like this:
`xpra attach ssh://sharon@192.168.10.91/0`

I do get the XTERM window
When I run in the XTERM window:
```
setxkbmap -query
```

I do see that he received from the client the keyboard layout for US and IL but it looks a little different than the client:
```
rules: evdev
model: dc105+inpc
layout: us,il,us
options: terminate: ctrl_alt_bksp
```

The difference I see is that in the model he gave a different name than what is in the local client
dc105+inpc instead of pc105+inet
And he didn't give "variant"

In any case, when I try to change the language, it continues to write in the XTERM window in English

But if I write manually in the XTERM window like this:
`setxkbmap -option grp:switch,alt_shift_toggle`

and pressing ALT-SHIFT it changes the language well
And changing the language using WIN+SPACE also works.


I haven't yet tested how it reacts if I connect via HTML5

**Reply by mayost** [reply]
Hello

I tried adding the parameters like this:
`xpra attach ssh://sharon@192.168.10.91/0 --keyboard-options=grp:switch,grp:alt_shift_toggle --keyboard-layouts=us,il`

When I write:
`setxkbmap -query`
I do get the result
```
rules: evdev
model: dc105+inpc
layout: us,il,us
options: grp:switch,alt_shift_toggle
```
But still when I try to change keyboard layout by ALT+SHIFT it doesn't work

Only when I register manually
```
setxkbmap -option ""
setxkbmap -option grp:switch,alt_shift_toggle
```
And then it works

Thanks

**Reply by mayost** [reply]
Hello

I performed another test on the new version v5.0.6-r3 in the Fedora 39 server
the problem is still there

Changing language at the client does not work

**Comment by mayost** [comment]
I noticed that if I am a client using other languages such as Arabic (which is also similar to Hebrew that is from right to left) or other languages such as French, German, etc.
It works very well without a problem
What I might think is that the problem with the Hebrew setting of the layout is IL and not HE, so maybe XPRA has a mistake that XPRA is trying on the server to set the Hebrew as HE and not as IL

This seems to me the direction to solve the problem because when I proactively change the layout with:
`setxkbmap -layout il`
It is indeed written in Hebrew

Thanks
